### PR TITLE
Parse fields in classes and generate correct class structs

### PR DESCRIPTION
**Pull request by eugene2k**
This pull request makes it possible for the generator to generate proper class structs, which is very useful for creating custom widgets for example. I haven't tested it much, mostly because I'm not very familiar with gtk-rs (or gtk+, for that matter), so if you see any problems with it, please point them out to me.

### CODE REVIEW

File: src/codegen/sys/lib_.rs [hunk 1]
```diff
@@ -72,6 +72,7 @@ fn generate_lib(w: &mut Write, env: &Env) -> Result<()>{
 
     try!(writeln!(w, "\n}}"));
 
+
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Why this empty line?

File: src/codegen/sys/lib_.rs [hunk 2]
```diff
@@ -232,10 +233,23 @@ fn prepare_enum_member_name(name: &str) -> String {
...
+        
+        let comment = if commented { "//" } else { "" };
+        if lines.is_empty() {
+            try!(writeln!(w, "{}#[repr(C)]\n{0}pub struct {}(c_void);\n", comment, klass.c_type));
+        }
+        else {
+            try!(writeln!(w, "{}#[repr(C)]\n{0}pub struct {} {{", comment, klass.c_type));
+
+            for line in lines {
+                try!(writeln!(w, "{}{}", comment, line));
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Isn't a comma missing in here?

File: src/codegen/sys/lib_.rs [hunk 3]
```diff
@@ -232,10 +233,23 @@ fn prepare_enum_member_name(name: &str) -> String {
...
-fn generate_classes_structs(w: &mut Write, classes: &[&Class]) -> Result<()> {
+fn generate_classes_structs(w: &mut Write, env: &Env, classes: &[&Class]) -> Result<()> {
     try!(writeln!(w, ""));
     for klass in classes {
-        try!(writeln!(w, "#[repr(C)]\npub struct {}(c_void);", klass.c_type));
+        let (lines, commented) = generate_fields(env, &klass.name, &klass.fields);
+        
+        let comment = if commented { "//" } else { "" };
+        if lines.is_empty() {
+            try!(writeln!(w, "{}#[repr(C)]\n{0}pub struct {}(c_void);\n", comment, klass.c_type));
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Could you use named parameters instead of `{}` and `{0}` please?

File: src/codegen/sys/lib_.rs [hunk 4]
```diff
@@ -232,10 +233,23 @@ fn prepare_enum_member_name(name: &str) -> String {
...
+        }
+        else {
+            try!(writeln!(w, "{}#[repr(C)]\n{0}pub struct {} {{", comment, klass.c_type));
```
Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Same comment as above.

**Comment by GuillaumeGomez** [comment]
Sorry for the delay.

cc @EPashkin @gkoz

File: src/codegen/sys/lib_.rs (hunk 1 above)
Comment (eugene2k) [inline src/codegen/sys/lib_.rs]: Actually, my whole file is different... wth...

Comment (eugene2k) [inline src/codegen/sys/lib_.rs]: Okay, this was stupid, but forgot to switch branches :)

File: src/codegen/sys/lib_.rs (hunk 2 above)
Comment (eugene2k) [inline src/codegen/sys/lib_.rs]: err... where?

Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Aren't you writing struct's fields? (It's totally possible that I misunderstood what you were doing here.)

Comment (eugene2k) [inline src/codegen/sys/lib_.rs]: Not quite, the fields are in `generate_fields`, which generates indented lines that end with a comma. I could move the comma out of generate fields, though.

Comment (eugene2k) [inline src/codegen/sys/lib_.rs]: Also, the original code wasn't written by me (I think @EPashkin wrote it), I simply modified it to work with classes as well as records, as I didn't see any reason it shouldn't.

Comment (GuillaumeGomez) [inline src/codegen/sys/lib_.rs]: Ok, so it confirms what I thought. My bad. :p

**Comment by GuillaumeGomez** [comment]
All good for me now. Waiting for @EPashkin and @gkoz.

**Comment by EPashkin** [comment]
Changes mady by this PR in sys crate: https://github.com/EPashkin/rust-gnome-sys/commit/be6f7c530095f271fb4ce90ed736e85572223d87
Not sure that we want show gtk internals, but for custom widget it may be needed (don't know how do create custom widget).

**Comment by EPashkin** [comment]
May be this need be separated as special mode.

**Comment by eugene2k** [comment]
Hmm... seeing as how the same types of fields are randomly private and public, I can force them all to be private, until someone actually raises an issue...

**Comment by gkoz** [comment]
I'm coming round to making all fields public as requested in https://github.com/gtk-rs/sys/issues/25. Given the advances in bindgen and apparent intention of the Rust project to integrate more C interoperability in 2017, it seems undesirable to diverge from C headers too much, and C doesn't have private fields.

**Comment by eugene2k** [comment]
> C doesn't have private fields
Which doesn't seem to be much of a problem for the GTK developers creating them...
/rant

**Comment by EPashkin** [comment]
@GuillaumeGomez, how you think, better make all fields public?
While `pub priv_: gpointer` look strange, it may be solve some problem with custom objects.

**Comment by EPashkin** [comment]
On other side, if we find, that all pub is wrong and unneeded changing it back it's be `breaking change`

**Comment by GuillaumeGomez** [comment]
I think that all FFI struct's fields should be public. Otherwise, it depends if it is an internal mechanism or not.

**Comment by EPashkin** [comment]
@GuillaumeGomez, seems it works, as you say.
@eugene2k, sorry for long wait, can you squash commits to one?

**Comment by EPashkin** [comment]
@GuillaumeGomez, about extra empty lines from https://github.com/gtk-rs/gir/blob/master/src/codegen/sys/lib_.rs#L245 etc: I think that better be replaced later with something like 
```
if(!interfaces.is_empty()) {
    try!(writeln!(w, "//Interfaces"));
}
```

**Comment by GuillaumeGomez** [comment]
@EPashkin: You mean `// Interfaces`, right? :p
@eugene2k: Just like @EPashkin said: please squash your commits into one and then I merge.

**Comment by eugene2k** [comment]
@EPashkin, @GuillaumeGomez 
I seem to have messed up my repo branches, so I'll resubmit the pull request in a moment

**Comment by GuillaumeGomez** [comment]
Sure, no worries.

**Pull request closed without merging**